### DISCUSSION TITLE
Is there anything to be gained with some kind of higher typeclass?

### DISCUSSION THREAD

**Discussion by oxinabox** [google-research/dex-lang#429] [Ideas]
Right now we have 3 monoid type classes:
 - `Mul` over `*` and `one`
 - `Add` over `+` and `zero`
 - `Monoid` over `<>` and `mempty` (possibly this should be called `FreeMonoid`?)
 

I don't have the background in functional programming to know if people do something cool with that.

Another example:
right now matmul `**` is defines explictly and only on `l=>m=>Float` matrixes,
and it uses `+` and `*` as the operations.
(well it actually uses `fsum` but that will at some point be `sum` i.e. `reduce +`)
One of the things you can do with matmul is:
If you have an adjancy `x` matrix, then you can do `x^p` (for ^ being matrix power) to get a matrix showing how many walks of length `p` there are between given vertexs, and you can sum them to get all the walks of length up to `p`.
But in principle the matmul operation can be generalized to any semiring.
Not just the one over `+` and `*`.
If you apply it to `|` and `&` (the boolean semiring), then `x^p` gives an answer to is there a walk bconnecting between vextexes of length `p`.
And if you do a sum to fixed point you get the connectivity matrix of the the graph.

idk if there is anothing that exposes this kind of abstraction, or of there is anything here to work with.

**Comment by apaszke** [comment]
Yeah, matmuls should just get parametrized by a `Semiring` typeclass. We just haven't implemented it yet.

**Reply by oxinabox** [reply]
Would a Semiring then be parameterized by a Monoid and and a Commutative Monoid?

**Comment by apaszke** [comment]
It's a choice we would have to make. While I do think that there's value in highlighting algebraic properties of data using those type classes, it is also quite easy to go too far and make the whole system too verbose and inaccessible for people without a proper algebra background. So I would treat that as a design/usability question, but `Semiring` seems important enough to warrant a type class.
